File: lib/linked_list.rb
class Linked_list
  attr_accessor :name

  def initialize
    @head = nil
    @tail = nil
  end

  def append(data)
    if @head.nil?
      @tail = Node.new
      @tail.value = data
      @head = @tail
    else
      temp_node = Node.new
      temp_node.value = data
      temp_node.next_node = nil
      @tail.next_node = temp_node
      @tail = temp_node
    end
  end

  def prepend(data)
    if @head.nil?
      @head = Node.new
      @head.value = data
      @tail = @head
    else
      temp_node = Node.new
      temp_node.value = data
      temp_node.next_node = @head
      @head = temp_node
    end
  end

  def size
    size = 0
    if @head.nil?
      "Current size is #{size}."
    elsif !@head.nil? && @head.next_node.nil?
      size += 1
      "Current size is #{size}."
    else
      size = 1
      current_node = @head
      loop do
        size += 1
        current_node = current_node.next_node
        break if current_node.next_node.nil?
      end
      size
    end
  end

  def head
    puts "Head of the list is #{@head.value}"
  end

  def tail
    puts "Tail of the list is #{@tail.value}"
  end

  def at(index)
    index_count = 0
    if @head.nil?
      'List has no values'
    elsif index == 0
      "Current value at index value of #{index} is #{@head.value}"
    else
      index_count += 1
      current_node = @head.next_node
      loop do
        current_node = current_node.next_node
        index_count += 1
        break if index_count == index || current_node.next_node.nil?
      end
      "Current value at index value of #{index} is #{current_node.value}"
    end
  end

  def pop
    current_node = @head
    loop do
      current_node = current_node.next_node
      break if current_node.next_node == @tail
    end
    @tail = nil
    current_node.next_node = nil
    @tail = current_node
  end

  def contains?(value)
    current_node = @head
    value_check = nil
    return true if current_node.value == value

    while current_node.next_node != nil # rubocop:disable Style/NonNilCheck
      current_node = current_node.next_node
      return true if current_node.value == value

      value_check = false

    end
    value_check
  end

  def find(value)
    index = 0
    current_node = @head
    return "#{current_node.value} found at index of #{index}" if current_node.value == value

    loop do
      current_node = current_node.next_node
      index += 1
      break if current_node.value == value || current_node.next_node.nil?
    end

    return nil if current_node.next_node.nil?

    "#{current_node.value} found at index of #{index}"
  end

  def to_s
    string = ''
    current_node = @head
    until current_node.nil?
      string += "( #{current_node.value} ) -> "
      current_node = current_node.next_node
    end
    string += 'nil'
    string
  end

  def insert_at(value, index)
    current_size = size
    target_index = 0
    current_node = @head
    prepend(value) if index.zero?
    append(value) if index == current_size
    puts 'Invalid index number' if index > current_size
    loop do
      current_node = current_node.next_node
      target_index += 1
      break if target_index == index - 1
    end
    new_node = Node.new
    new_node.value = value
    new_node.next_node = current_node.next_node
    current_node.next_node = new_node
  end

  def remove_at(index) # doesnt work w/index of 1
    if index.zero?
      @head = @head.next_node
    elsif index == 1
      node_to_remove = @head.next_node
      @head.next_node = node_to_remove.next_node
    else
      current_size = size
      target_index = 0
      current_node = @head
      puts 'Invalid index number' if index > current_size
      loop do
        current_node = current_node.next_node
        target_index += 1
        break if target_index == index - 1
      end
      next_node = current_node.next_node
      current_node.next_node = next_node.next_node
    end
  end
end

File: lib/linked_list.js
import { Node } from "./node.js";

export { LinkedList };

class LinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
  }

  append(value) {
    const node = new Node(value);

    if (this.head === null) {
      this.head = node;
      this.tail = node;
    } else {
      this.tail.nextNode = node;
      this.tail = node;
    }
  }

  prepend(value) {
    const node = new Node(value);

    if (this.head === null) {
      this.head = node;
      this.tail = node;
    } else {
      node.nextNode = this.head;
      this.head = node;
    }
  }

  size() {
    if (this.head === null) return 0;

    let size = 1;
    let node = this.head;

    while (node.nextNode !== null) {
      node = node.nextNode;
      size += 1;
    }

    return size;
  }

  headCheck() {
    if (this.head === null) {
      return undefined;
    } else {
      return this.head.value;
    }
  }

  tailCheck() {
    if (this.tail === null) {
      return undefined;
    } else {
      return this.tail.value;
    }
  }

  at(index) {
    let currentNode = this.head;

    if (index > this.size()) {
      return undefined;
    } else {
      for (let i = 0; i <= index; i++) {
        if (i === index) {
          return currentNode.value;
        } else {
          currentNode = currentNode.nextNode;
        }
      }
    }
  }

  nodeAt(index) {
    let currentNode = this.head;

    if (index > this.size()) {
      return undefined;
    } else {
      for (let i = 0; i <= index; i++) {
        if (i === index) {
          return currentNode;
        } else {
          currentNode = currentNode.nextNode;
        }
      }
    }
  }

  pop() {
    if (this.head === null) {
      return undefined;
    } else {
      this.head = this.head.nextNode;
    }
  }

  contains(value) {
    let node = this.head;
    while (node.nextNode !== null) {
      if (node.value === value) {
        return true;
      } else {
        return false;
      }
      node = node.nextNode;
    }
  }

  findIndex(value) {
    let counter = 0;
    let node = this.head;
    while (node !== null) {
      if (node.value === value) {
        return counter;
      }
      counter += 1;
      node = node.nextNode;
    }
    return -1;
  }

  toString() {
    let string = ""
    let node = this.head;
    while (node !== null) {
      string = string + ` ( ${node.value} ) ->`
      node = node.nextNode;
    }
    string += " null";
    return string;
  }

  insertAt(index, ...values) {
    // create a node for each of the added values
    const newNodes = []
    values.forEach((value) => {
      const node = new Node(value);
      newNodes.push(node);
    })
    // define nextNode for all newNodes except last one
    newNodes.forEach((node) => {
      if (newNodes.indexOf(node) <= newNodes.length - 2) {
        node.nextNode = newNodes[newNodes.indexOf(node) + 1]
      }
    })

    // define the insertion point and insert new nodes
    if (index < 0 || index > this.size()) {
      return RangeError
    } else if (index === 0) {
      const shiftNode = this.head;
      newNodes[newNodes.length - 1].nextNode = shiftNode;
      this.head = newNodes[0];
    } else if (index === this.size) {
      const currentTail = this.tail
      currentTail.nextNode = newNodes[0]

      this.tail = newNodes[newNodes.length - 1];
    } else {
      const insertionNode = this.nodeAt(index - 1);
      const shiftNode = this.nodeAt(index)

      insertionNode.nextNode = newNodes[0]
      newNodes[newNodes.length - 1].nextNode = shiftNode;
    }
  }

  removeAt(index) {
    if (index < 0 || index > this.size()) {
      return RangeError
    } else if (index === 0) {
      const newHead = this.head.nextNode
      this.head = newHead
    } else if (index >= 0) {
      const previousNode = this.nodeAt(index - 1);
      const nodeToRemove = this.nodeAt(index);

      if (nodeToRemove === this.tail) { this.tail = previousNode }

      previousNode.nextNode = nodeToRemove.nextNode;

      if (previousNode === this.tail) {
        this.tail = newNodes[newNodes.length - 1];
      }
    }
  }

  test() {
    const node = this.head
    console.log(node.nextNode.nextNode)
    console.log(node.nextNode.value)
  }

}
